src/main/java/com/bigchaindb/builders/BigchainDbTransactionBuilder.java

```diff
@@ -371,12 +371,26 @@ public IBuild build(EdDSAPublicKey publicKey) {
          */
         private void sign(EdDSAPrivateKey privateKey)
                 throws InvalidKeyException, SignatureException, NoSuchAlgorithmException {
-
             String temp = this.transaction.toHashInput();
-            log.debug("TO BE HASHED ---->\n" + temp + "\n<");
             JsonObject transactionJObject = DriverUtils.makeSelfSortingGson(temp);
 
-            byte[] sha3Hash = DriverUtils.getSha3HashRaw(transactionJObject.toString().getBytes());
+            byte[] sha3Hash;
+            if (Operations.TRANSFER.name().equals(this.transaction.getOperation())) {
+                // it's a transfer operation: make sure to update the hash pre-image with
+                // the fulfilling transaction IDs and output indexes
+                StringBuilder preimage = new StringBuilder(transactionJObject.toString());
+                for (Input in : this.transaction.getInputs()) {
```

Comment:
the fix itself: if it is a `TRANSFER` operation, do the following:
- `for` each `Input`...
  - concatenate `transactionId` and `String.valueOf(outputIndex)`
  - add to end of hash preimage

all before calculating the hash and signing.